### ISSUE TITLE
Add indicator to all external links via CSS

### ISSUE THREAD

**Issue by mkllnk** [openfoodfoundation/openfoodnetwork#13030]
## Description

When a user clicks on a link, they should know if it is an internal or external link. External links are usually opened in a new tab. And some parts of our code are already adding an icon. But adding the icon in HTML can make translations containing HTML more complicated.

We suggest to use a CSS rule based on the target:
```css
a[target="_blank"]::after {
  /*show the icon.. */
}
```

We currently display it like this:

```html
<i class="icon-external-link"></i>
```

This came up in:

* https://github.com/openfoodfoundation/openfoodnetwork/pull/13011#pullrequestreview-2493921844

## Acceptance Criteria & Tests

1. All links that currently have an indicator icon, should still have it.
2. Other links with `target="_blank"` should have the icon, including in user-defined text like shop messages.
3. The current rich text editors should still work as before (with or without displaying the icon, doesn't matter).
4. Links in the menus should not have the icon. Even if they go to an external page like `https://about.openfoodnetwork.org.au` it's considered internal.
5. You will need to browse through lots of parts of the app to ensure that nothing breaks and everything looks correct.

**Comment by mkllnk** [comment]
Thank you for working through this! Those are some really good points. I would like to have a closer look at our use cases to make a good decision on this.

> Due to various websites using links inconsistently users are confused whether the icon means new tab or external link

In our case, we only open a new tab for external links and all external links open a new tab. So whatever the user thinks it means, it's right. Some users may still not be confident because they agonise about the effect. I'm probably a user like that. But I would think the same about a link without symbol. So I always open new tabs myself when that's important to me.

The inconsistency is not a strong argument for me. The question is more if the indicator adds any value. We probably need to examine each case individually.

> Possible exceptions would be:
> 
>     linking from a secure/logged in site to somewhere out of the system.
>     the new tab's content only adds information and is not part of the flow you want users to follow.

In the example which triggered this issue both of these exceptions apply. A user is logged in and is about to activate an integration. They can follow a link to gain more information but they leave the current workflow in that case.

You could argue though that they can simple use the back button instead of closing the tab. And there are probably more people knowing about the back button of the browser than about the concepts of tabs.

I personally use a lot of tabs and get annoyed when a link, for example in Github, opens within the window and I lose the exact view I had before. But I also know how to open new tabs myself. I'm not the target group here.

> Add (opens in new tab) right at the end of the links e.g. Producers (opens in new tab). 

I like that idea. Would you include it in the label of the link? For example:
```html
<a href="https://example.net" target="_blank">Example documentation (opens in new tab)</a>
```
In that case, we can easily include it in translations and don't have any icon code in the way.

> Add both rel="noreferrer noopener" as well as target="_blank" for security

I totally agree. This should be added.

> we should think about incorporating a linking strategy/guide for both content and code.

Our current strategy has been to mark links to external sites. This mostly happens in user-generated content like enterprise and product descriptions. But there's also the Connected Apps page where each connected app can have links for more information.

Adding _(opens in new tab)_ to user-generated content is a bit more challenging but possible.

But maybe we should challenge this practice generally. I know that there are some recommendations to just not open new tabs by default and let the user decide / do the work. Would that be your preference as well?

**Comment by krisztin** [comment]
Thank you for providing more context. This is btw one of those things where I know the best practice but if you really need convincing, to get project specific data we should be asking actual users instead of us making assumptions 😅️. I've floated the idea of some user testing in the #design-circle Slack but I'm not a User Researcher so I have no idea how to lead on / go about it, I usually just analyse research data.

### Should you ever open new tabs?
Tbh, I just went with guidelines not opening new tabs, never really questioning why. I've now looked into it more going through W3's WCAG Techniques and [checking on NNG](https://www.nngroup.com/articles/new-browser-windows-and-tabs/) (THE source for the best **research-based** UX).

TL;DR: it depends but 99% of the cases it should just open in the same tab and if a user wants a new tab they can do it themselves.

#### Good to know
- The back button is indeed more well known and used.
- New tabs are especially cumbersome on mobile devices

#### Cases when you CAN open new tabs/windows (in a bit more detail)
1. Linking from logged in state - "The user is logged into a secured area of a site, and following a link to a page outside of the secured area would terminate the user's logon. In this case opening external links in an external window allows the user to access such references while keeping their login active in the original window." [General Technique (G200) from W3](https://www.w3.org/WAI/WCAG22/Techniques/general/G200)
2. Linking from a form - in order to keep the data the user already filled in on a form.
3. Comparing information - user needs multiple tabs to be able to compare, for example, products
4. Combining information - e.g. In one window you have a tax form, in the other some reference material to aid the filling OR your example of getting tasks done on Github where you might want the PR and issue in separate tabs for instance (I have a feeling Github didn't really do user testing and just went with the "users will sort it out" method)

#### Example link
Yes, sorry my example was rubbish. You are correct, just need to add the `rel` attributes as well so a new tab link should be:

```html
<a href="https://example.net" target="_blank" rel="noreferrer noopener">Example documentation (opens in new tab)</a>
```

If you have a list of external links, you could do the following:

```html
<h2>The following links open in a new tab</h2>
<ul>
  <li>
    <a href="https://example1.net" target="_blank" rel="noreferrer noopener">Example 1 documentation <span class="visually-hidden">(opens in new tab)</span></a>
  </li>
  <li>
    <a href="https://example2.net" target="_blank" rel="noreferrer noopener">Example 2 documentation <span class="visually-hidden">(opens in new tab)</span></a>
  </li>
  <li>
    <a href="https://example3.net" target="_blank" rel="noreferrer noopener">Example 3 documentation <span class="visually-hidden">(opens in new tab)</span></a>
  </li>
</ul>
```

#### My suggestion
Based on best practice and generic UX research we can decide on a case-by-case basis with most links not requiring a `target`

### Icons

> The inconsistency is not a strong argument for me. The question is more if the indicator adds any value. We probably need to examine each case individually.

The conclusion at GOV.UK was that it does not add any value. The value would be the user knowing what will happen but they don't. Users also don't care about whether a link is external or internal.

**Comment by mkllnk** [comment]
Well, I'm convinced. 😄 

I propose these actions to align with best practices and will ask the team what they think about it:

* We do not open links in new tabs by default. It's annoying on mobile and users like to choose themselves.
* Remove icons/indicators for new tabs/external sites because users don't find them useful.
* Add `rel="noreferrer"` to all external links.

With that last point, it's annoying that browsers don't do that automatically. Some privacy browsers probably do that but break some websites. I chose to add only `norferrer` because that should imply `noopener` as well:

> omit the [Referer](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referer) header and otherwise leak no referrer information — and additionally to behave as if the [noopener](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/noopener) keyword were also specified.
https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/noreferrer

And if that's all true then you don't need to do it with opening a new tab either:

> Note: Setting target="_blank" on <a>, <area> and <form> elements implicitly provides the same rel behavior as setting rel="noopener" which does not set window.opener.
https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/noopener

**Comment by RachL** [comment]
Hello @krisztin and nice to meet you! I have a few questions to clearly understand:

> Remove icons/indicators for new tabs/external sites because users don't find them useful.

Do we have an example of what type of icons we are looking at?

Is it easy to list the places where the behavior will change? I'm thinking especially on places we have filters, I know in some cases keeping the filters the user has selected is not working, so if we don't open in new tab and use the back button it can be annoying to have to reselect everything.

**Comment by mkllnk** [comment]
> Do we have an example of what type of icons we are looking at?

Yes, we started this issue discussing connected apps. Here's the page:

![Image](https://github.com/user-attachments/assets/59a8f26d-ba5f-4af6-80b2-1517cc769ced)

There are two external links with these icons, like this:

![Image](https://github.com/user-attachments/assets/9d2eebd5-00ac-4132-a6e4-f8ef0cd9db50)

In these cases I agree that the icon doesn't add any value. And it's not necessary to open a new tab. After navigating to the external resources, you can use the browser's back button to go back to OFN. And you won't lose any information on the Connected Apps page.

**Comment by mkllnk** [comment]
> Is it easy to list the places where the behavior will change?

Connected Apps is one place.

Then we use the same icon (`icon-external-link`) in the product upload for the CSV template downloads. It doesn't open a new tab and is not an external link. So that's confusing!

We also use it for some links to the user guide. It would be good to go through all cases. There may be some within a form that should still open in a new tab. And then that could be indicated with a text label instead.

The last use of that icon is in the admin layout, the store link at the top right opens in a new tab. I don't think that's necessary either. It's actually quite confusing that if you go to Administration then it opens a new tab and if you then click on Store then it opens a new tab as well. So you can navigate back to the place you were but you then have three open tabs instead of one.

**Comment by krisztin** [comment]
Hi @RachL 👋️ Nice to meet you too!

Just to clarify we are talking about two new issues:
1. **Remove all external link icons**
Fingers crossed the below are the only affected pages/instances
2. **Review (and if necessary update) all links opening in new tabs**
My search for `_blank` turned up 710 results. This is just an approximate, there could be more. So this might be a bit bigger project that could do with further breaking down. Not sure how.

### External icon instances
I'm still getting familiar with the codebase but my search for `icon-external-link` (it's a fairly easy search in my code editor software btw) turned up 17 instances of it being used. Not sure if this is helpful, I am probably missing something. What do you think @mkllnk, does this look right?

```markdown
app/views/admin/product_import/_upload_sidebar.html.haml:
  4    %a.download{href: '/product_list_template.csv'}
  5:     %i.icon-external-link
  6      = t('admin.product_import.index.product_list_template')

  8    %a.download{href: '/inventory_template.csv'}
  9:     %i.icon-external-link
  10      = t('admin.product_import.index.inventory_template')

app/views/admin/shared/_user_guide_link.html.haml:
  1: = button_link_to t('.user_guide'), ContentConfig.user_guide_link, icon: 'icon-external-link', target: '_blank'

app/views/spree/layouts/admin/_login_nav.html.haml:
  12      %li
  13:       %i.icon-external-link
  14        = link_to t(".header.store"), main_app.root_path, target: "_blank"

config/locales/en_CA.yml:
  1371:             link_label_html: "Learn more about this research project <i class=\"icon-external-link\"></i>"
  1385:             link_label_html: "Learn more about Discover Regenerative <i class=\"icon-external-link\"></i>"
  1396:             link_label_html: "VINE <i class=\"icon-external-link\"></i>"

config/locales/en_FR.yml:
  1371:             link_label_html: "Learn more about this research project <i class=\"icon-external-link\"></i>"
  1385:             link_label_html: "Learn more about Discover Regenerative <i class=\"icon-external-link\"></i>"
  1396:             link_label_html: "VINE <i class=\"icon-external-link\"></i>"

config/locales/en.yml:
  1441:             link_label_html: "Learn more about this research project <i class=\"icon-external-link\"></i>"
  1455:             link_label_html: "Learn more about Discover Regenerative <i class=\"icon-external-link\"></i>"
  1466:             link_label_html: "VINE <i class=\"icon-external-link\"></i>"

config/locales/fr_CA.yml:
  1371:             link_label_html: "En savoir plus à propos de ce projet de recherche <i class=\"icon-external-link\"></i>"
  1385:             link_label_html: "En savoir plus sur Discover Regenerative  <i class=\"icon-external-link\"></i>"
  1396:             link_label_html: "VINE <i class=\"icon-external-link\"></i>"

config/locales/fr.yml:
  1369:             link_label_html: "En savoir plus à propos de ce projet de recherche <i class=\"icon-external-link\"></i>"
  1383:             link_label_html: "En savoir plus sur Discover Regenerative <i class=\"icon-external-link\"></i>"
  1394:             link_label_html: "Editeur de bons de réduction VINE <i class=\"icon-external-link\"></i>"
  
 ```

**Comment by mkllnk** [comment]
You can ignore all locales except en.yml.

**Comment by RachL** [comment]
thank you both! Removing icons sounds logical :+1: My concerns was more regarding should we completely stop opening in a new page ornot.

> the store link at the top right opens in a new tab. I don't think that's necessary either. It's actually quite confusing that if you go to Administration then it opens a new tab and if you then click on Store then it opens a new tab as well. So you can navigate back to the place you were but you then have three open tabs instead of one.

Perhaps, but this is assuming no one is reaching the admin page without going through the shopfront (through a bookmark for example). I will try to check what Matomo can tell us (ping @filipefurtad0  FYI).
But having store + admin on 2 different tab is quite handy (you update on one and refresh on the other :) )

> So this might be a bit bigger project that could do with further breaking down. Not sure how.

Indeed, maybe we should do one header menu and its sub pages at a time? I think there are menus like dashboard or list of customers, where i'm quite sure there is no issues. However others where we can search and filter lists require maybe a last check.

**Comment by krisztin** [comment]
@RachL, this 'new tab or not' conundrum sounds like the perfect candidate for some user testing just so that we could learn how actual users interact with the system. If this hasn't been done before ofc. I've floated the idea of some testing in the #design-circle on Slack (but at that point had no clue where we could start) and Bethan Phillips mentioned he could rustle up a few users if needed.

**Comment by RachL** [comment]
hey @krisztin I'm really sorry for the late answer here. I wanted to take the time to answer properly and then got caught in all the things...

I think it's awesome if you can sync with Bethan to talk to users. Let me know if you need me. Also I'm available for a general call if you want, as an opportunity to learn more about each others. Don't hesitate to DM on slack if you fancy
